### ISSUE TITLE
Phased P2P rollout

### ISSUE THREAD

**Issue by lyulka** [parallelchain-io/parallelchain-protocol#19]
## Basic idea

There’s an old P2P network, and there’s a new P2P network.

On startup, a Fullnode can be configured to either:
1. Connect to both the old network and the new network, or
2. Connect to the new network only.

If it connects to both networks, it sends a message to, and receives a message from, both networks. However, applications of the network must only receive the message once.

### Segregating the two networks

Should the two networks use:
* The same Kademlia overlay network, and two different topics? Or,
* Two different Kademlia overlay networks.

### Message deduplication

* How do we ID messages?
* What kind of data structure do we use to remember whether a peer’s received a message before? What are the invalidation rules?

## Proposed design

The old version of P2P and the new version will be separate Kademlia overlay networks. The old version (v0.4) keeps the existing protocol name “pchain_p2p/v1.0.0”, and the new version (v0.5) will have the protocol name “parallelchain_mainnet_p2p/v0.5”. 

Replica implementations must be written to support connecting to:
* Both the v0.4 and the v0.5 networks.
* Only the v0.5 network.

### Message deduplication

If a replica connects to both networks, all messages it sends must be sent on both the old network and the new network, with the same pseudorandom message ID. Receiving replicas cache this message ID to deduplicate messages: only one of the messages should be passed to the P2P application. 

Since message ID collisions can happen incidentally even with pseudorandom message IDs, and messages that collide only incidentally should be processed, caches should have an aggressive invalidation policy. 

An example implementation of a cache with an invalidation policy has a HashMap, and a FIFO queue. After the HashMap has reached a configurable limit on its number of entries, every insertion to the HashMap will pop the oldest value from the queue and remove it from the HashMap.

### Minimizing boilerplate

Instead of implementing both the v0.4 and v0.5 P2P protocols in a single version of pchain-network, have individual versions of pchain-network implement a single version of the P2P protocol, and have Fullnode import two versions: one implementing P2P v0.4, and one implementing P2P v0.5.

### Questions for `pchain-network` implementation

How do we override message IDs to send a message in both v0.4 and v0.5 networks with the same message IDs? It’s not immediately obvious how to do so using the [message_id](https://docs.rs/libp2p/latest/libp2p/gossipsub/struct.Config.html#method.message_id) function. Some multi-threaded trickery may be necessary.

## Phased rollout schedule

To avoid requiring a coordinated shutdown and restart of all replicas to effect the change from P2P v0.4 to v0.5, the rollout will be "phased".

Fullnode software implementing protocol v0.5 will support both P2P v0.4 and v0.5. A configuration variable will be made available that lets operators choose between using both P2P v0.4 and v0.5, or both. E.g.:

```rust
enum P2PVersion {
    Both,
    V05Only,
}
```

A fullnode configured to use both versions of P2P will send and process GossipSub messages in both v0.4 and v0.5 networks.
With this option, phased rollout of P2P can proceed as follows:
1. On each of 4 designated dates, replicas representing 20% of validator power will shut down and start with new binaries, with `P2PVersion::Both` selected.
2. On a single, final date, the last 20% of validator power will shut down and start with new binaries, now with `P2PVersion::V05Only` selected.
3. Then, on 4 further designated dates, replicas restarted in step 1 restart again, this time with `P2PVersion::V05Only` selected.

The assignment of validators to restart dates will be done with consultation with the community.

**Comment by lyulka** [comment]
@Alvinhon please look at the design on GitHub and offer thoughts, in particular about how we could override message IDs to send a message in both v0.4 and v0.5 networks with the same message IDs

**Comment by AlvinHon** [comment]
There are two approaches:
- > have individual versions of pchain-network implement a single version of the P2P protocol
- > both the v0.4 and v0.5 P2P protocols in a single version of pchain-network

The message_id function is useful in the second approach because messages are delivered in the same network. There is a cache to store the message keyed by message_id. If two different message with same message_id are received, only one of the messages can be proceeded. So in this way, messages with different versions can be de-duplicate by defining a message_id function such that they have same message_id.

In the first approach, there are two networks isolating with each other. So, there are two caches. It is not obvious how we can synchronize them just by defining a message_id function. In this approach, instead overriding message_id function in libp2p, the application of pchain-network should maintain a cache for the messages received from different p2p networks by itself. Overall, there are two running networks simultaneously and Fullnode will be responsible for implementing part of the network messaging protocol.

**Comment by lyulka** [comment]
Another design choice is between having:
1. Peers connect to two different libp2p networks, one for the old version of P2P, and one for the new version.
2. Peers subscribe to two different GossipSub topics, one for the old version of P2P, and one for the new version.

This design chooses the first option since it is more general. In particular, if in the future we want to switch to a new, not backwards compatible version of libp2p GossipSub, the second approach will not work, but the first approach will.

A concern with the first option is the extra resources that a peer needs to use to connect to two different libp2p networks. However, peers will only connect to two different libp2p networks in a short transition period, and will only connect with one network after that.

**Comment by AlvinHon** [comment]
## Fullnode (P2PVersion::Both)

- Fullnode imports two network configurations for old and new version of pchain-network
- Fullnode runs to pchain_network
- Fullnode sends message in both v0.4 and v0.5 network
- Fullnode implements a cache for message de-duplication


## Message

Suppose the message in v0.4 is the `data` with type `Vec<u8>`. The message in v0.5 adds a field `group_id` at the beginning:

```rust
struct {
	group_id: Option<u64>,
	data: Vec<u8>
}
```

The field `group_id` is used for specifying group of equivalent messages which are in different versions. The u64 value is set to the `sequence_number` of the message in the lowest version (v0.4).

Fullnode (P2PVersion::Both) sends the message in the lowest version (v0.4) first, and then sends message in higher version.

For example, to deliver message in both v0.4 and v0.5:
1. Sends message in v0.4 where `sequence_number` = `x`
2. Sends message in v0.5 where `group_id` = `Some(x)`

If Fullnode supports only v0.5 network (P2PVersion::V05Only), the field `group_id` is set to be `None`.

## Cache

Fullnode (P2PVersion::Both) implements a LRU time cache to store the received message from both v0.4 and v0.5 networks:
- The message in v0.4 is inserted into cache by key `source + sequence_number`. 
- The message in v0.5 is inserted into cache by key 
    - `source + sequence_number` if `group_id` is `None`
    - `source + group_id.unwrap()` if `group_id` is `Some`

When a received message has already inserted into the cache (having same key), it should be ignored. In other words, message in old version is supposed to be proceed first.

Reminder: there is no mechanism to punish the node who intentionally sends duplicate messages (i.e. send messages in v0.4 and v0.5 with `group_id` = `None`, etc).

**Comment by lyulka** [comment]
@AlvinHon a few comments and a question:

- Instead of `data` being `Vec<u8>`, it should be an enum of all possible messages on the P2P network as specified in the protocol.
- If `group_id` is `None`, we know that the sender is only connected to the v0.5 network, so why insert a key into the LRU cache?
- The name `group_id` can be more explicit, maybe: `v04_sequence_number`.
- The sequence number of a message is decided inside a `pchain_network` and inside libp2p. How do we communicate this to the other `pchain_network` to be set as `group_id`?

**Comment by AlvinHon** [comment]
> Instead of data being Vec<u8>, it should be an enum of all possible messages on the P2P network as specified in the protocol.

Yes. The exact format of `data` should follow the specification of the protocol. e.g. `data` in p2p network v0.5 follows the protocol v0.5.

> If group_id is None, we know that the sender is only connected to the v0.5 network, so why insert a key into the LRU cache?

It is for generalization. Maybe the same de-duplicate mechanism can be applied to the upgrade from v0.5 and v0.6. If it is true, by storing the `source + sequence_number` allows de-duplicating the v0.6 message (which specifies `group_id` to be sequence number of v0.5 message). 

> The name group_id can be more explicit, maybe: v04_sequence_number.

It is related to the previous answer.

> The sequence number of a message is decided inside a pchain_network and inside libp2p. How do we communicate this to the other pchain_network to be set as group_id?

The [publish](https://docs.rs/libp2p-gossipsub/latest/libp2p_gossipsub/struct.Behaviour.html#method.publish) function in libp2p gossipsub will return `MessageId` which allows us to get the `sequence_number`. And then the remaining part is to return it to the host application. One way is that `pchain_network` can setup a channel for sending this information to the host application which then sends message with `group_id` to other pchain_network.

**Comment by lyulka** [comment]
An alternative way to pair v0.4 and v0.5 messages: having both v0.4 and v0.5 message share the same message ID. 

We can do so by overriding `message_id_fn` in the config of the v0.5 P2P implementation with a closure that holds a `Receiver<MessageID>`. A procedure for publishing a message to both v0.4 and v0.5 P2P networks would be to:
1. Call `publish` on the v0.4 implementation, getting back a message ID.
2. Send the message ID down the channel to the v0.5 `message_id_fn`.
3. Call `publish` on the v0.5 implementation. In the call, the implementation calls the overriden `message_id_fn`, which in turns receives the message ID in the channel and returns it to the implementation to set the ID for the message in the v0.5 network.

This reduces complexity in the specification while making the implementation only slightly more complex. What do you think @AlvinHon?

**Comment by AlvinHon** [comment]
I feel like the message receiver side is missing something for computing the "`MessageID`" of the received v0.5 message. The input of the function `message_id_fn` only consists of `source`, `sequence_number`, `data` and `topic`. Maybe I need more clarifications.

**Comment by lyulka** [comment]
Agreed with @AlvinHon that this even simpler design for the v0.4 and v0.5 "network adapter" is better:
> When I receive a message from someone on the v0.5 p2p network, I mark this peer as having upgraded to v0.5, I will then ignore all messages from it in the v0.4 p2p network. Again this cache should have a ttl just in case the operator of the replica rolls back the binary to v0.4 for some reason.

**Comment by lyulka** [comment]
Merged into #8.
